### DISCUSSION TITLE
Support for TRACE sub-element within INTRODUCTION

### DISCUSSION THREAD

**Discussion by rcd09** [cogu/autosar#120] [Ideas]
I am working with 0.5.3 reading in a platform base type arxml and it is causing reader.read_file to throw an exception: 
`PlatformBase_Types.arxml(88): Unprocessed element <TRACE>
TypeError: MultiLanguageReferrable.__init__() got an unexpected keyword argument 'introduction'`

The offending ARXML:

>            <SW-BASE-TYPE>
>               <SHORT-NAME>sint8</SHORT-NAME>
>               <LONG-NAME>
>                 <L-4 L="EN">signed integer 8bit</L-4>
>               </LONG-NAME>
>               <CATEGORY>FIXED_LENGTH</CATEGORY>
>               <INTRODUCTION>
>                 <TRACE>
>                   <SHORT-NAME>SWS_Platform_00016</SHORT-NAME>
>                   <CATEGORY>SPECIFICATION_ITEM</CATEGORY>
>                   <P>
>                     <L-1 L="EN">This standard AUTOSAR type shall be 8 bit signed</L-1>
>                   </P>
>                 </TRACE>
>               </INTRODUCTION>
>               <BASE-TYPE-SIZE>8</BASE-TYPE-SIZE>
>               <BASE-TYPE-ENCODING>2C</BASE-TYPE-ENCODING>
>               <MEM-ALIGNMENT>8</MEM-ALIGNMENT>
>             </SW-BASE-TYPE>

Question 1: I have searched for a good guide to an AUTOSAR ARXML specification without a lot of success. The closest I've found is AUTOSAR_TPS_SoftwareComponentTemplate.pdf and its tough to find what I need there.  What's a good doc to look at when researching ARXML syntax?

Question 2: When I take out the INTRODUCTION elements with the TRACE subs all is good - file is read in just fine.  Is TRACE a left over from an older ARXML version?  Maybe an extension used by our SIP supplier? Or something that needs to be supported here?  I'd be happy to add it if it should be supported.

Thank you!

**Comment by cogu** [comment]
Hi @rcd09,

Yes, it's indeed very hard to find the specification. On [autosar.org](autosar.org) you find it hidden away inside a file called `AUTOSAR_FO_MMOD_XMLSchema.zip`. Inside it you will find the Schema file for the ARXML. 

The reference file I use for all my development is named `AUTOSAR_00051.xsd` which is part of R22. At first I used a fancy XML tool to render the schema for me graphically which was very helpful to understand how it's constructed.

In the XSD there is a valid element called TRACE which I have not yet implemented. 
After a quick glance I noticed that it's definition doesn't follow normal convention The tricky part is the "ref=" part which see to indicate that it includes a whole other type inside it. The current Python code does not support such construction as far as I know.

This would need further investigation and most likely need moderate refactoring of the code base to add that support. I would not recommend trying to add it by yourself but you are of course welcome to try. I struggled a lot with implementing the documentation elements due to its recursive definition.
